Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs

```diff
@@ -90,63 +90,75 @@ public void Stop()
-                    walletTip = this.chain.GetBlock(fork.HashBlock);
-                    this.logger.LogTrace("Wallet tip set to '{0}/{1}'.", walletTip.HashBlock, walletTip.Height);
+                    this.logger.LogTrace("Wallet tip set to '{0}'.", this.walletTip);
                 }
 
-                ChainedBlock incomingBlock = this.chain.GetBlock(block.GetHash());
-                if (incomingBlock.Height > this.walletTip.Height)
+                // The new tip can be ahead or behind the wallet
+                // If the new tip is ahead we try to bring the wallet up to the new tip.
+                // If the new tip is behind we just check the wallet and the tip are in the same chain.
+            
+                if (newTip.Height > this.walletTip.Height)
                 {
-                    CancellationToken token = this.nodeLifetime.ApplicationStopping;
+                    ChainedBlock findTip = newTip.FindAncestorOrSelf(this.walletTip.HashBlock);
+                    if (findTip == null)
+                    {
+                        this.logger.LogTrace("(-)[NEW_TIP_AHEAD_NOT_IN_WALLET]");
+                        return;
+                    }
+
+                    var token = this.nodeLifetime.ApplicationStopping;
+                    this.logger.LogTrace("Wallet tip '{0}/{1}' is behind the new tip '{2}/{3}'.", this.walletTip.HashBlock, this.walletTip.Height, newTip.HashBlock, newTip.HashBlock);
 
-                    // The wallet is falling behind we need to catch up.
                     ChainedBlock next = this.walletTip;
-                    while (next != incomingBlock)
+                    while (next != newTip)
                     {
-                        token.ThrowIfCancellationRequested();
-
-                        // While the wallet is catching up the entire node will wait
-                        // if a wallet recovers to a date in the past. Consensus 
-                        // will stop till the wallet is up to date.
+                        // While the wallet is catching up the entire node will wait.
+                        // If a wallet is recovered to a date in the past. Consensus will stop till the wallet is up to date.
 
                         // TODO: This code should be replaced with a different approach
-                        // similar to BlockStore. The wallet should be standalone and not depend on consensus.
+                        // Similar to BlockStore the wallet should be standalone and not depend on consensus.
                         // The block should be put in a queue and pushed to the wallet in an async way.
                         // If the wallet is behind it will just read blocks from store (or download in case of a pruned node).
 
-                        next = this.chain.GetBlock(next.Height + 1);
+                        token.ThrowIfCancellationRequested();
+
+                        next = newTip.GetAncestor(next.Height + 1);
```

Comment:
as mentioned somewhere, i would prefer here to have a list of such `next` blocks instead of looking for one again and again

Reply:
It wont matter much as the store can anyway just take one block at the time.
Also this whole code might change later so I wont bother with this refactor right now.

Reply:
it won't, it just feels bad from theoretical point; you won't have here very long differences in height between newTip and next, so O(N^2) complexity should be OK, but from theoretical point, this is just ugly

but exactly as you say - no need to do anything about it since this will be changed (soon)

so we are good with this, just a general note